Спринт 16

KaktysTV1 opened the pull request:
Выполнена задача спринта 16

Commit:
Выполнена задача спринта 16

Helvetii951 commented on Traker/AddingHabits/Category/CategoriesViewController.swift:
```diff
@@ -1,24 +1,13 @@
-    var categories: [String] = []
     var selectedCategory: String = ""
     
     // MARK: - Private properties:
+    private let viewModel: CategoryViewModel
     private let categoryStore = TrackerCategoryStore.shared
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом и как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

Helvetii951 commented on Traker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,119 @@
+
+class OnboardingViewController: UIPageViewController {
+    // MARK: - Private Properties:
+    private lazy var pages: [UIViewController] = {
+        let firstPage = FirstPageViewController()
+        let secondPage = SecondPageViewController()
+        
+        return [firstPage, secondPage]
+    }()
```
`FirstPageViewController` и `SecondPageViewController` отличаются только текстом и картинкой. В этом случае лучше написать один универсальный. Так кода будет меньше и он будет гибче.

Helvetii951 commented on Traker/Setings/SceneDelegate.swift:
```diff
@@ -17,8 +17,11 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         window?.makeKeyAndVisible()
+        
+        DaysTransformer.register()
+        ColorTransformer.register()
```
Стартовый контроллер устанавливается и в `AppDelegate`, и в `SceneDelegate`. Очень стоит оставить в одном месте. Желательно в `SceneDelegate`.

Helvetii951 commented on Traker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,119 @@
+            fatalError("Не удалось инициализировать AppDelegate")
+        }
+        appDelegate.window?.rootViewController = TabBarViewController()
+    }
```
Лучше забрать у контроллера онбординга обязанность навигации и отдать. ее тому, кто создал контроллер(в нашем случае это `SplashViewController`). Передаем событие нажатия кнопки через, например, замыкание, а `SplashViewController` уже пусть сам решает что делать.

Helvetii951 commented on Traker/Onboarding/SpashViewController/FirstLaunchStorage.swift:
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+final class FirstLaunchStorage {
+    // MARK: - Properties:
+    static let shared = FirstLaunchStorage()
+    var wasLaunchedBefore: Bool {
+        get {
+            userDefaults.bool(forKey: "firstLaunch")
+        }
```
`FirstLaunchStorage` слишком узкий нэйминг. С ростом приложения таких флагов становится все больше и больше. Всегда желательно иметь один класс, из которого их можно вытащить. Подошло бы, например, `UserSettingStorage`.

Helvetii951 commented on Traker/Onboarding/SpashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,49 @@
+import UIKit
+
+final class SplashViewController: UIViewController {
+    // MARK: - Private Properties:
+    private let firstLaunchStorage = FirstLaunchStorage.shared
+    private var wasLaunchedBefore: Bool = FirstLaunchStorage().wasLaunchedBefore
+    
```
Здесь используется и статический экземпляр `shared` и создается новый экземпляр `FirstLaunchStorage()`. Лучше использовать `shared` раз уж он есть. А еще лучше будет сделать `FirstLaunchStorage` [синглтоном](https://theswiftdev.com/swift-singleton-design-pattern/).

Helvetii951 commented on Traker/Onboarding/Pages/FirstPageViewController.swift:
```diff
@@ -0,0 +1,69 @@
+import UIKit
+
+class FirstPageViewController: UIViewController {
+    // MARK: - Private Properties:
```
Вряд ли класс будет наследоваться, поэтому стоит пометить его `final`. Преимущества:

1. Никто не сможет по ошибке сделать подкласс.
2. Присвоение классу статуса `final` указывает компилятору Swift, что функции следует вызывать напрямую (статическая диспетчеризация), а не искать функцию в таблице методов (динамическая диспетчеризация). Это снижает затраты на вызов функций и повышает производительность. Подробнее об этом можно прочитать [в блоге разработчика Swift](https://developer.apple.com/swift/blog/?id=27).

Commit:
Исправлены критические замечания
(К сожалению, у меня не остается время на выполнение рекомендаций. После сдачи 17 спринта с жестким дедлайном, обязательно займусь рекомендациями)
